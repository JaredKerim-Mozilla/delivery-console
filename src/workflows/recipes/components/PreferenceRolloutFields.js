/* eslint-disable react/jsx-boolean-value */
import { Row, Col, Button, Input, InputNumber, Select } from 'antd';
import { Map, List } from 'immutable';
import PropTypes from 'prop-types';
import React from 'react';

import FormItem from 'console/components/forms/FormItem';
import SwitchBox from 'console/components/forms/SwitchBox';
import { connectFormProps } from 'console/utils/forms';

const DEFAULT_VALUES = {
  string: '',
  boolean: false,
  integer: 0,
};

export const serializePreferenceRows = prefsList => {
  if (!prefsList) {
    return List([]);
  }
  let _nextId = 0;
  return prefsList.map(pref => {
    const value = pref.get('value');
    let type;
    if (typeof value === 'string') {
      type = 'string';
    } else if (typeof value === 'boolean') {
      type = 'boolean';
    } else if (typeof value === 'number') {
      type = 'integer';
    } else {
      throw new Error(`Unrecognized preference type ${typeof value} (${value})`);
    }
    return {
      name: pref.get('preferenceName'),
      type,
      value,
      id: _nextId++,
    };
  });
};

export const deserializePreferenceRows = prefsList => {
  return List(
    // The reason for the filter is to remove any form argument items that are null
    // or undefined which can happen when deleting a <FormItem> object.
    prefsList.filter(row => !!row).map(row => {
      return {
        preferenceName: row.name,
        value: row.value,
      };
    }),
  );
};

@connectFormProps
class PreferenceRolloutFields extends React.Component {
  state = {
    rows: serializePreferenceRows(this.props.recipeArguments.get('preferences')),
  };

  static propTypes = {
    disabled: PropTypes.bool,
    form: PropTypes.object.isRequired,
    recipeArguments: PropTypes.instanceOf(Map),
  };

  static defaultProps = {
    disabled: false,
    recipeArguments: new Map(),
  };

  // Use a locally increasing number too to avoid the case were a new ID is one that
  // it's already been.
  _nextId = 0;

  handleClickAddNewRow = () => {
    this.setState(state => {
      // Default to the same type as the last row.
      // Useful if you add a bunch of booleans one after the other.
      const type = state.rows.length ? state.rows[state.rows.length - 1].type : 'string';

      // The nextId needs never be a number we've used before.
      // XXX This is getting complicated. Maybe just use Math.random()/
      const ids = [...state.rows.values()].map(r => r.id);
      const nextId = 1 + (ids.length ? Math.max(...ids) : 0) + this._nextId++;

      return {
        rows: state.rows.push({
          name: '',
          type: type,
          value: DEFAULT_VALUES[type],
          id: nextId,
        }),
      };
    });
  };

  changeRow = row => {
    this.setState(state =>
      state.rows.map(r => {
        if (r.id === row.id) {
          return row;
        }
        return r;
      }),
    );
  };

  removeRow = row => {
    // Don't allow this if there are currently only 1 rows.
    if (this.state.rows.size <= 1) {
      throw new Error('There must be at least one row left.');
    }

    this.setState(state => {
      return {
        rows: state.rows.filter(r => r.id !== row.id),
      };
    });
  };

  render() {
    const { disabled, recipeArguments } = this.props;

    return (
      <div>
        <Row>
          <p className="action-info">XXX - Come up with a better line here.</p>
          <Col sm={24} md={11}>
            <FormItem
              label="Slug"
              name="arguments.slug"
              initialValue={recipeArguments.get('slug', '')}
            >
              <Input disabled={disabled} />
            </FormItem>
          </Col>
        </Row>

        <p className="action-info">Preferences</p>

        {this.state.rows.map(row => (
          <RowField
            row={row}
            defaultValues={DEFAULT_VALUES}
            disabled={disabled}
            key={row.id}
            disableRemove={this.state.rows.size <= 1}
            onRowChange={row => {
              this.changeRow(row);
            }}
            onRowRemove={row => {
              this.removeRow(row);
            }}
          />
        ))}

        <Button
          disabled={disabled}
          type="default"
          icon="plus"
          onClick={this.handleClickAddNewRow}
          id="add-row"
        >
          Add Row
        </Button>
      </div>
    );
  }
}

export default PreferenceRolloutFields;

export class RowField extends React.PureComponent {
  state = {
    name: this.props.row.name,
    type: this.props.row.type,
    value: this.props.row.value,
  };

  static propTypes = {
    row: PropTypes.object.isRequired,
    defaultValues: PropTypes.object,
    disabled: PropTypes.bool,
    disableRemove: PropTypes.bool,
  };

  static defaultProps = {
    defaultValues: DEFAULT_VALUES,
    disabled: false,
    disableRemove: false,
  };

  // XXX Not a great name.
  _reportRow = () => {
    this.props.onRowChange({
      name: this.state.name,
      type: this.state.type,
      value: this.state.value,
    });
  };

  handleNameChange = name => {
    this.setState(() => {
      return { name: name.trim() };
    }, this._reportRow);
  };

  handleTypeChange = type => {
    const { defaultValues } = this.props;
    if (!defaultValues.hasOwnProperty(type)) {
      throw new Error(`Unrecognized value type for '${type}'`);
    }
    this.setState(
      {
        type,
        value: defaultValues[type],
      },
      this._reportRow,
    );
  };

  handleValueChange = value => {
    this.setState({ value }, this._reportRow);
  };

  render() {
    // The reason for passing the `i` is so that we can have a unique thing that
    // makes the `name` on each <FormItem> different and unique.
    const { disabled, row, disableRemove } = this.props;
    const { id } = row;
    const { name, type, value } = this.state;

    let valueInput;
    if (type === 'string') {
      valueInput = (
        <Input
          disabled={disabled}
          onChange={event => this.handleValueChange(event.target.value)}
        />
      );
    } else if (type === 'boolean') {
      valueInput = <SwitchBox disabled={disabled} onChange={this.handleValueChange} />;
    } else if (type === 'integer') {
      valueInput = <InputNumber disabled={disabled} onChange={this.handleValueChange} />;
    } else {
      throw new Error(`Unrecognized value type (${type})`);
    }

    return (
      <Row gutter={8} align="bottom">
        <Col span={6}>
          <FormItem label="Name" name={`arguments.preferences.${id}.name`} initialValue={name}>
            <Input
              disabled={disabled}
              onChange={event => this.handleNameChange(event.target.value)}
            />
          </FormItem>
        </Col>
        <Col span={4}>
          <FormItem label="Type" name={`arguments.preferences.${id}.type`} initialValue={type}>
            <Select disabled={disabled} onChange={this.handleTypeChange}>
              <Select.Option value="string">String</Select.Option>
              <Select.Option value="boolean">Boolean</Select.Option>
              <Select.Option value="integer">Integer</Select.Option>
            </Select>
          </FormItem>
        </Col>
        <Col span={10}>
          <FormItem label="Value" name={`arguments.preferences.${id}.value`} initialValue={value}>
            {valueInput}
          </FormItem>
        </Col>
        <Col span={4} style={{ textAlign: 'right' }}>
          <div className="ant-form-item-no-colon">
            <div className="ant-form-item-label">
              <label> </label>
            </div>
            <div>
              <Button
                disabled={disabled || disableRemove}
                type="default"
                title={disableRemove ? 'There must be at least 1 preference' : null}
                icon="close"
                onClick={() => this.props.onRowRemove(row)}
                id={`${id}.remove-row`}
              >
                Remove Row
              </Button>
            </div>
          </div>
        </Col>
      </Row>
    );
  }
}
