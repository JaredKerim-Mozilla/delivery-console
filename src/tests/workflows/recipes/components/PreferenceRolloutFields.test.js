import { Map, List } from 'immutable';
import PreferenceRolloutFields, {
  RowField,
  serializePreferenceRows,
  deserializePreferenceRows,
} from 'console/workflows/recipes/components/PreferenceRolloutFields';

describe('<PreferenceRolloutFields>', () => {
  it('should work', () => {
    const props = {
      disabled: false,
      form: {},
      recipeArguments: new Map(),
    };
    const wrapper = () => shallow(<PreferenceRolloutFields {...props} />);
    expect(wrapper).not.toThrow();
  });

  // Hopefully we can remove this test and not have to test this deep implementation detail.
  // Ideally we should just be talking directly to the parent component (PreferenceRolloutFields).
  describe('<RowField>', () => {
    const props = {
      id: 0,
      name: 'name',
      type: 'string',
      value: 'Peter',
    };

    it('should work', () => {
      const wrapper = () => shallow(<RowField {...props} />);
      expect(wrapper).not.toThrow();
    });
  });

  // Hopefully, when we have better test coverage we shouldn't need these specific
  // tests but instead we render the whole component with or without a list of
  // default preferences. Then, ideally a more encompassing test should test all of
  // these specifics "automatically".
  describe('serializing and deserializing preferences', () => {
    it('serialize a prefsList to a list of objects', () => {
      const list = [
        Map({ preferenceName: 'name', value: 'Peter' }),
        Map({ preferenceName: 'swede', value: true }),
        Map({ preferenceName: 'kids', value: 2 }),
      ];
      const prefsList = new List(list);
      const rowList = serializePreferenceRows(prefsList);
      expect(rowList).toBeImmutableList();
      expect(rowList.get(0)).toMatchObject({
        name: 'name',
        type: 'string',
        value: 'Peter',
      });
      expect(rowList.get(1)).toMatchObject({
        name: 'swede',
        type: 'boolean',
        value: true,
      });
      expect(rowList.get(2)).toMatchObject({
        name: 'kids',
        type: 'integer',
        value: 2,
      });
      expect(rowList.size).toEqual(3);
    });

    it('should throw error on unrecognized type', () => {
      const list = [Map({ preferenceName: 'names', value: ['Peter'] })];
      const prefsList = new List(list);
      expect(() => {
        serializePreferenceRows(prefsList);
      }).toThrow('Unrecognized preference type object (Peter)');
    });

    it('should serialize a falsy prefsList to a list of 0 objects', () => {
      const rowList = serializePreferenceRows(undefined);
      expect(rowList).toBeImmutableList();
      expect(rowList.size).toEqual(0);
    });

    it('should deserialize a List of rows into a proper preferences List', () => {
      const rows = List([
        { name: 'name', type: 'string', value: 'Peter' },
        { name: 'kids', type: 'integer', value: 2 },
        { name: 'swede', type: 'boolean', value: true },
      ]);
      const prefsList = deserializePreferenceRows(rows);
      expect(prefsList).toBeImmutableList();
      expect(prefsList.get(0)).toMatchObject({
        preferenceName: 'name',
        value: 'Peter',
      });
      expect(prefsList.get(1)).toMatchObject({
        preferenceName: 'kids',
        value: 2,
      });
      expect(prefsList.get(2)).toMatchObject({
        preferenceName: 'swede',
        value: true,
      });
      expect(prefsList.size).toEqual(3);
    });
  });
});
