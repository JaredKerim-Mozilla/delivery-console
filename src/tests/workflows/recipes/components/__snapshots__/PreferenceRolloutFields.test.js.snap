// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PreferenceRolloutFields> <RowField> should work 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RowField
    defaultValues={
      Object {
        "boolean": false,
        "integer": 0,
        "string": "",
      }
    }
    disableRemove={false}
    disabled={false}
    row={
      Object {
        "name": "name",
        "type": "string",
        "value": "Peter",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "align": "bottom",
      "children": Array [
        <Col
          span={6}
        >
          <Wrapper
            initialValue="name"
            label="Name"
            name="arguments.preferences.undefined.name"
          >
            <Input
              disabled={false}
              onChange={[Function]}
              prefixCls="ant-input"
              type="text"
            />
          </Wrapper>
        </Col>,
        <Col
          span={4}
        >
          <Wrapper
            initialValue="string"
            label="Type"
            name="arguments.preferences.undefined.type"
          >
            <Select
              choiceTransitionName="zoom"
              disabled={false}
              onChange={[Function]}
              prefixCls="ant-select"
              showSearch={false}
              transitionName="slide-up"
            >
              <Option
                value="string"
              >
                String
              </Option>
              <Option
                value="boolean"
              >
                Boolean
              </Option>
              <Option
                value="integer"
              >
                Integer
              </Option>
            </Select>
          </Wrapper>
        </Col>,
        <Col
          span={10}
        >
          <Wrapper
            initialValue="Peter"
            label="Value"
            name="arguments.preferences.undefined.value"
          >
            <Input
              disabled={false}
              onChange={[Function]}
              prefixCls="ant-input"
              type="text"
            />
          </Wrapper>
        </Col>,
        <Col
          span={4}
          style={
            Object {
              "textAlign": "right",
            }
          }
        >
          <div
            className="ant-form-item-no-colon"
          >
            <div
              className="ant-form-item-label"
            >
              <label>
                 
              </label>
            </div>
            <div>
              <Button
                block={false}
                disabled={false}
                ghost={false}
                icon="close"
                id="undefined.remove-row"
                loading={false}
                onClick={[Function]}
                prefixCls="ant-btn"
                title={null}
                type="default"
              >
                Remove Row
              </Button>
            </div>
          </div>
        </Col>,
      ],
      "gutter": 8,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <Wrapper
            initialValue="name"
            label="Name"
            name="arguments.preferences.undefined.name"
          >
            <Input
              disabled={false}
              onChange={[Function]}
              prefixCls="ant-input"
              type="text"
            />
          </Wrapper>,
          "span": 6,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <Input
              disabled={false}
              onChange={[Function]}
              prefixCls="ant-input"
              type="text"
            />,
            "initialValue": "name",
            "label": "Name",
            "name": "arguments.preferences.undefined.name",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "disabled": false,
              "onChange": [Function],
              "prefixCls": "ant-input",
              "type": "text",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <Wrapper
            initialValue="string"
            label="Type"
            name="arguments.preferences.undefined.type"
          >
            <Select
              choiceTransitionName="zoom"
              disabled={false}
              onChange={[Function]}
              prefixCls="ant-select"
              showSearch={false}
              transitionName="slide-up"
            >
              <Option
                value="string"
              >
                String
              </Option>
              <Option
                value="boolean"
              >
                Boolean
              </Option>
              <Option
                value="integer"
              >
                Integer
              </Option>
            </Select>
          </Wrapper>,
          "span": 4,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <Select
              choiceTransitionName="zoom"
              disabled={false}
              onChange={[Function]}
              prefixCls="ant-select"
              showSearch={false}
              transitionName="slide-up"
            >
              <Option
                value="string"
              >
                String
              </Option>
              <Option
                value="boolean"
              >
                Boolean
              </Option>
              <Option
                value="integer"
              >
                Integer
              </Option>
            </Select>,
            "initialValue": "string",
            "label": "Type",
            "name": "arguments.preferences.undefined.type",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <Option
                  value="string"
                >
                  String
                </Option>,
                <Option
                  value="boolean"
                >
                  Boolean
                </Option>,
                <Option
                  value="integer"
                >
                  Integer
                </Option>,
              ],
              "choiceTransitionName": "zoom",
              "disabled": false,
              "onChange": [Function],
              "prefixCls": "ant-select",
              "showSearch": false,
              "transitionName": "slide-up",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "String",
                  "value": "string",
                },
                "ref": null,
                "rendered": "String",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "Boolean",
                  "value": "boolean",
                },
                "ref": null,
                "rendered": "Boolean",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "Integer",
                  "value": "integer",
                },
                "ref": null,
                "rendered": "Integer",
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <Wrapper
            initialValue="Peter"
            label="Value"
            name="arguments.preferences.undefined.value"
          >
            <Input
              disabled={false}
              onChange={[Function]}
              prefixCls="ant-input"
              type="text"
            />
          </Wrapper>,
          "span": 10,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <Input
              disabled={false}
              onChange={[Function]}
              prefixCls="ant-input"
              type="text"
            />,
            "initialValue": "Peter",
            "label": "Value",
            "name": "arguments.preferences.undefined.value",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "disabled": false,
              "onChange": [Function],
              "prefixCls": "ant-input",
              "type": "text",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <div
            className="ant-form-item-no-colon"
          >
            <div
              className="ant-form-item-label"
            >
              <label>
                 
              </label>
            </div>
            <div>
              <Button
                block={false}
                disabled={false}
                ghost={false}
                icon="close"
                id="undefined.remove-row"
                loading={false}
                onClick={[Function]}
                prefixCls="ant-btn"
                title={null}
                type="default"
              >
                Remove Row
              </Button>
            </div>
          </div>,
          "span": 4,
          "style": Object {
            "textAlign": "right",
          },
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="ant-form-item-label"
              >
                <label>
                   
                </label>
              </div>,
              <div>
                <Button
                  block={false}
                  disabled={false}
                  ghost={false}
                  icon="close"
                  id="undefined.remove-row"
                  loading={false}
                  onClick={[Function]}
                  prefixCls="ant-btn"
                  title={null}
                  type="default"
                >
                  Remove Row
                </Button>
              </div>,
            ],
            "className": "ant-form-item-no-colon",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <label>
                   
                </label>,
                "className": "ant-form-item-label",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": " ",
                },
                "ref": null,
                "rendered": " ",
                "type": "label",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <Button
                  block={false}
                  disabled={false}
                  ghost={false}
                  icon="close"
                  id="undefined.remove-row"
                  loading={false}
                  onClick={[Function]}
                  prefixCls="ant-btn"
                  title={null}
                  type="default"
                >
                  Remove Row
                </Button>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "block": false,
                  "children": "Remove Row",
                  "disabled": false,
                  "ghost": false,
                  "icon": "close",
                  "id": "undefined.remove-row",
                  "loading": false,
                  "onClick": [Function],
                  "prefixCls": "ant-btn",
                  "title": null,
                  "type": "default",
                },
                "ref": null,
                "rendered": "Remove Row",
                "type": [Function],
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "align": "bottom",
        "children": Array [
          <Col
            span={6}
          >
            <Wrapper
              initialValue="name"
              label="Name"
              name="arguments.preferences.undefined.name"
            >
              <Input
                disabled={false}
                onChange={[Function]}
                prefixCls="ant-input"
                type="text"
              />
            </Wrapper>
          </Col>,
          <Col
            span={4}
          >
            <Wrapper
              initialValue="string"
              label="Type"
              name="arguments.preferences.undefined.type"
            >
              <Select
                choiceTransitionName="zoom"
                disabled={false}
                onChange={[Function]}
                prefixCls="ant-select"
                showSearch={false}
                transitionName="slide-up"
              >
                <Option
                  value="string"
                >
                  String
                </Option>
                <Option
                  value="boolean"
                >
                  Boolean
                </Option>
                <Option
                  value="integer"
                >
                  Integer
                </Option>
              </Select>
            </Wrapper>
          </Col>,
          <Col
            span={10}
          >
            <Wrapper
              initialValue="Peter"
              label="Value"
              name="arguments.preferences.undefined.value"
            >
              <Input
                disabled={false}
                onChange={[Function]}
                prefixCls="ant-input"
                type="text"
              />
            </Wrapper>
          </Col>,
          <Col
            span={4}
            style={
              Object {
                "textAlign": "right",
              }
            }
          >
            <div
              className="ant-form-item-no-colon"
            >
              <div
                className="ant-form-item-label"
              >
                <label>
                   
                </label>
              </div>
              <div>
                <Button
                  block={false}
                  disabled={false}
                  ghost={false}
                  icon="close"
                  id="undefined.remove-row"
                  loading={false}
                  onClick={[Function]}
                  prefixCls="ant-btn"
                  title={null}
                  type="default"
                >
                  Remove Row
                </Button>
              </div>
            </div>
          </Col>,
        ],
        "gutter": 8,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <Wrapper
              initialValue="name"
              label="Name"
              name="arguments.preferences.undefined.name"
            >
              <Input
                disabled={false}
                onChange={[Function]}
                prefixCls="ant-input"
                type="text"
              />
            </Wrapper>,
            "span": 6,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <Input
                disabled={false}
                onChange={[Function]}
                prefixCls="ant-input"
                type="text"
              />,
              "initialValue": "name",
              "label": "Name",
              "name": "arguments.preferences.undefined.name",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "disabled": false,
                "onChange": [Function],
                "prefixCls": "ant-input",
                "type": "text",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <Wrapper
              initialValue="string"
              label="Type"
              name="arguments.preferences.undefined.type"
            >
              <Select
                choiceTransitionName="zoom"
                disabled={false}
                onChange={[Function]}
                prefixCls="ant-select"
                showSearch={false}
                transitionName="slide-up"
              >
                <Option
                  value="string"
                >
                  String
                </Option>
                <Option
                  value="boolean"
                >
                  Boolean
                </Option>
                <Option
                  value="integer"
                >
                  Integer
                </Option>
              </Select>
            </Wrapper>,
            "span": 4,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <Select
                choiceTransitionName="zoom"
                disabled={false}
                onChange={[Function]}
                prefixCls="ant-select"
                showSearch={false}
                transitionName="slide-up"
              >
                <Option
                  value="string"
                >
                  String
                </Option>
                <Option
                  value="boolean"
                >
                  Boolean
                </Option>
                <Option
                  value="integer"
                >
                  Integer
                </Option>
              </Select>,
              "initialValue": "string",
              "label": "Type",
              "name": "arguments.preferences.undefined.type",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  <Option
                    value="string"
                  >
                    String
                  </Option>,
                  <Option
                    value="boolean"
                  >
                    Boolean
                  </Option>,
                  <Option
                    value="integer"
                  >
                    Integer
                  </Option>,
                ],
                "choiceTransitionName": "zoom",
                "disabled": false,
                "onChange": [Function],
                "prefixCls": "ant-select",
                "showSearch": false,
                "transitionName": "slide-up",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "children": "String",
                    "value": "string",
                  },
                  "ref": null,
                  "rendered": "String",
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "children": "Boolean",
                    "value": "boolean",
                  },
                  "ref": null,
                  "rendered": "Boolean",
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "children": "Integer",
                    "value": "integer",
                  },
                  "ref": null,
                  "rendered": "Integer",
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <Wrapper
              initialValue="Peter"
              label="Value"
              name="arguments.preferences.undefined.value"
            >
              <Input
                disabled={false}
                onChange={[Function]}
                prefixCls="ant-input"
                type="text"
              />
            </Wrapper>,
            "span": 10,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <Input
                disabled={false}
                onChange={[Function]}
                prefixCls="ant-input"
                type="text"
              />,
              "initialValue": "Peter",
              "label": "Value",
              "name": "arguments.preferences.undefined.value",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "disabled": false,
                "onChange": [Function],
                "prefixCls": "ant-input",
                "type": "text",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <div
              className="ant-form-item-no-colon"
            >
              <div
                className="ant-form-item-label"
              >
                <label>
                   
                </label>
              </div>
              <div>
                <Button
                  block={false}
                  disabled={false}
                  ghost={false}
                  icon="close"
                  id="undefined.remove-row"
                  loading={false}
                  onClick={[Function]}
                  prefixCls="ant-btn"
                  title={null}
                  type="default"
                >
                  Remove Row
                </Button>
              </div>
            </div>,
            "span": 4,
            "style": Object {
              "textAlign": "right",
            },
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <div
                  className="ant-form-item-label"
                >
                  <label>
                     
                  </label>
                </div>,
                <div>
                  <Button
                    block={false}
                    disabled={false}
                    ghost={false}
                    icon="close"
                    id="undefined.remove-row"
                    loading={false}
                    onClick={[Function]}
                    prefixCls="ant-btn"
                    title={null}
                    type="default"
                  >
                    Remove Row
                  </Button>
                </div>,
              ],
              "className": "ant-form-item-no-colon",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <label>
                     
                  </label>,
                  "className": "ant-form-item-label",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": " ",
                  },
                  "ref": null,
                  "rendered": " ",
                  "type": "label",
                },
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <Button
                    block={false}
                    disabled={false}
                    ghost={false}
                    icon="close"
                    id="undefined.remove-row"
                    loading={false}
                    onClick={[Function]}
                    prefixCls="ant-btn"
                    title={null}
                    type="default"
                  >
                    Remove Row
                  </Button>,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "block": false,
                    "children": "Remove Row",
                    "disabled": false,
                    "ghost": false,
                    "icon": "close",
                    "id": "undefined.remove-row",
                    "loading": false,
                    "onClick": [Function],
                    "prefixCls": "ant-btn",
                    "title": null,
                    "type": "default",
                  },
                  "ref": null,
                  "rendered": "Remove Row",
                  "type": [Function],
                },
                "type": "div",
              },
            ],
            "type": "div",
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`<PreferenceRolloutFields> should work 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Wrapper
    disabled={false}
    form={Object {}}
    recipeArguments={Immutable.Map {}}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "disabled": false,
      "form": Object {},
      "formErrors": undefined,
      "recipeArguments": Immutable.Map {},
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "disabled": false,
        "form": Object {},
        "formErrors": undefined,
        "recipeArguments": Immutable.Map {},
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
